### DISCUSSION TITLE
How to create Program dependency graph in ghidra?

### DISCUSSION THREAD

**Discussion by SimaArasteh** [NationalSecurityAgency/ghidra#7030] [Q&A]
Hello, 

I know that PDG is a combination of CDG and DDG. I could create a data dependency graph using pcodeAST in ghidra but I do not know how to create control dependency and PDG. I've read some scripts but it was about creating control flow which is different from CDG. How can I create CDG with ghidra? Which API I should look into?

Moreover is there an alternative approach to create DDG in ghidra?

Thanks

**Comment by d-millar** [comment]
Although probably not exactly what you want, you might look at the GraphAST.java script in Ghidra/Features/Decompiler/ghidra_scripts (and maybe the GraphASTAndFlow script).  Despite its name, it is closer to a DDG than an AST (my opinion) and exposes some interesting bits of the API.  Am sure others will correct me on this, but it’s at least a starting point.

**Comment by SimaArasteh** [comment]
Thanks I have looked at that but I am not sure if it is a control flow or control dependency! Since these graphs are different I want to make sure that working with pcodeAST gives the data and control dependency?!

**Comment by d-millar** [comment]
Gotcha - I’m not sure the dominance trees used by the decompiler are exposed through the API.  Chris is definitely the guy to weigh in on this.  Will ping him tomorrow if he doesn’t jump on it immediately.

**Comment by SimaArasteh** [comment]
Hi. sorry to ping you again but since I found ghidra's ddg accurate and I would like to use it for my research I would really appreciate it if you can help me with this question. Thanks a lot

**Comment by d-millar** [comment]
Mentioned this to @caheckman this morning - he's going to see whether the dominator trees are exposed.

**Comment by caheckman** [comment]
Hi.  The decompiler internally calculates a standard dominance graph, but this is not the inverse dominance graph required for a formal control dependence graph.  Ghidra does have a class for computing inverse dominance on a generic directed graph, ChkPostDominanceAlgorithm.  It requires a graph object implementing the GDirectedGraph interface, so you would need to build a control-flow graph version of this.  Look at the method makeSelectionFromVertex in the class SelectByScopedFlowPlugin in the Ghidra source code for an example calculating the inverse/post dominance graph from the control-flow graph of a function.

**Comment by SimaArasteh** [comment]
Thanks a lot for your response. For data dependence graph (DDG), In PcodeAST , I used varnode.getDescendants() method to get access to varnodes that data depend on each other I just wanna make sure that for DDG I am looking on the right API. 

an One more question. Is there anyway to get access to CDG and DDG without using decompilation ?! thanks


Sima

**Comment by caheckman** [comment]
The Decompiler is currently the only code calculating data dependence (technically Static Single Assignment form).  You can use the class BasicBlockModel to compute control-flow instead of using the decompiler.

**Comment by SimaArasteh** [comment]
Thanks alot for your support. just to make sure, for DDG using varnodes in pcodeAST is a correct way to calculate DDG. right? thanks
